### ISSUE TITLE
During prediction, would the result be stochastic?

### ISSUE THREAD

**Issue by yqNLP** [AlCorreia/HILAM#1]
Since it is related to a sample operation(the attention vector), so every time we sample, we would get a different attention vector, so the accuracy would change accordingly?

**Comment by AlCorreia** [comment]
Yes, that is correct. The feature selector is essentially a mask that we sample from some distribution associated with relative feature importance. Thus, the mask is indeed stochastic and may vary among different runs with the same input. The actual impact on the prediction and, per consequence, on the accuracy depends on how robust the model is, but indeed there is no guarantee that you would get the same output every time.
I experimented with two alternatives to this behaviour. The simplest one is to replace the sampling with an argmax in test time. That way you take the most likely set of features every time and the prediction is no longer stochastic.
The other one is to increase the number of samples to render the model more robust. There is even an attribute ``num_samples`` in the ``model`` class that I used to specify the number of masks to sample for each example. This is more elegant but also considerably increases the computational cost of the model. As I did not observe a great improvement in the predictions, I kept ``num_samples=1`` in most of my experiments.
